src/utils/lazy.cpp

```diff
@@ -0,0 +1,113 @@
+#include "../../shared/utils/lazy.hpp"
+
+#include <cassert>
+#include <condition_variable>
+#include <mutex>
+
+constexpr std::uintptr_t UNINIT = 0x0;
```

Comment:
I'm personally adverse to having `constexpr` constants, instead preferring `enum` or `enum class` which results in the same result but is in a more "segmented" location, though this again doesn't really matter too much.